api/helpers.go

```diff
@@ -91,3 +107,196 @@ func paginationToRequest[T any](rows []T, dbPageSize int32, cursor string, goFor
+	if err != nil {
+		return nil, "", nil, err
+	}
+	db.Lock()
+	defer db.Unlock()
+	if err := db.Del(bID); err != nil {
+		return nil, "", nil, err
+	}
+	return root, uri, dump, nil
+}
+
+// InnerCensusID generates a unique identifier by concatenating the BlockNumber, StrategyID,
+// and a numerical representation of the Anonymous flag from a CreateCensusRequest struct.
+// The BlockNumber and StrategyID are concatenated as they are, and the Anonymous flag is
+// represented as 1 for true and 0 for false. This concatenated string is then converted
+// to a uint64 to create a unique identifier.
+func InnerCensusID(blockNumber, strategyID uint64, anonymous bool) uint64 {
+	// Convert the boolean to a uint64: 1 for true, 0 for false
+	var anonymousUint uint64
+	if anonymous {
+		anonymousUint = 1
+	}
+	// Concatenate the three values as strings
+	concatenated := fmt.Sprintf("%d%d%d", blockNumber, strategyID, anonymousUint)
+	// Convert the concatenated string back to a uint64
+	result, err := strconv.ParseUint(concatenated, 10, 64)
+	if err != nil {
+		panic(err)
+	}
+	if result > math.MaxInt64 {
+		panic(err)
+	}
+	return result
+}
+
+// CalculateStrategyHolders function returns the holders of a strategy and the
+// total weight of the census. It also returns the total block number of the
+// census, which is the sum of the strategy block number or the last block
+// number of every token chain id. To calculate the census holders, it uses the
+// supplied predicate to filter the token holders using a lexer and evaluator.
+// The evaluator uses the strategy operators to evaluate the predicate which
+// uses the database queries to get the token holders and their balances, and
+// combines them.
+func CalculateStrategyHolders(ctx context.Context, qdb *queries.Queries, w3p state.Web3Providers,
+	id uint64, predicate string,
+) (map[common.Address]*big.Int, *big.Int, uint64, error) {
+	// init some variables to get computed in the following steps
+	censusWeight := new(big.Int)
+	strategyHolders := map[common.Address]*big.Int{}
+	// parse the predicate
+	lx := lexer.NewLexer(strategyoperators.ValidOperatorsTags)
+	validPredicate, err := lx.Parse(predicate)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	// get strategy tokens from the database
+	strategyTokens, err := qdb.TokensByStrategyID(ctx, id)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, nil, 0, err
+		}
+		return nil, nil, 0, err
+	}
+	// any census strategy is identified by id created from the concatenation of
+	// the block number, the strategy id and the anonymous flag. The creation of
+	// censuses on specific block is not supported yet, so we need to get the
+	// last block of every token chain id to sum them and get the total block
+	// number, used to create the census id.
+	totalTokensBlockNumber := uint64(0)
+	for _, token := range strategyTokens {
+		w3uri, exists := w3p[token.ChainID]
+		if !exists {
+			return nil, nil, 0, err
+		}
+		w3 := state.Web3{}
```

Comment:
I think this "getLastBlockForChainID" should be abstracted and part of w3p

Reply:
This will be wrapped with the implementation of Web3 tokens of [this interface](https://github.com/vocdoni/census3/blob/main/service/holders_provider.go)